File: dp_tabulation/pickup_deliveries/pickup_deliveries.cpp
#include <cmath>
using namespace std;

int count_valid_options(int total_orders) // LeetCode Q.1359.
{
    long long valid_pickups = 1, sequence_len = 2; // Base case at n = 1: (P1, D1).

    long long modulo = pow(10, 9) + 7;
    for (int order = 2; order <= total_orders; order++)
    {
        long long combos = (1 + sequence_len + 1) * (sequence_len + 1) / 2;
        combos %= modulo;
        valid_pickups *= combos;
        valid_pickups %= modulo;
        sequence_len += 2; // Update sequence len for next order amount.
    }

    return valid_pickups;
}

File: dp_tabulation/pickup_deliveries/pickup_deliveries.py

def count_valid_options(total_orders: int) -> int:  # LeetCode Q.1359.
    valid_pickups = 1  # Base case at n = 1: (P1, D1).

    modulo = 10 ** 9 + 7
    for order in range(2, total_orders + 1):
        last_order_seq_len = 2 * (order - 1)
        combos = (1 + last_order_seq_len + 1) * (last_order_seq_len + 1) // 2
        valid_pickups *= combos
        valid_pickups %= modulo

    return valid_pickups
